### PR TITLE
Add iso_download_pve feature to additional_iso_files

### PR DESCRIPTION
**Pull request by ELT4N1N**
This PR adds the `iso_download_pve` feature to `additional_iso_files`. This is helpful for iso files that are a on the bigger side and are often used for multiple VMs, like the virtio drivers. The flag can only be used together with iso_url.
The code is based on the code from the iso module and shares it with it. 

It was tested on PVE  7.4-3.

### CODE REVIEW

**Comment by nywilken** [comment]
Thanks for opening up this change we will review as soon as we get a chance.

File: builder/proxmox/common/config.go
Comment (nywilken) [inline builder/proxmox/common/config.go:369]: It look like you want to append to errs here. 
```suggestion
			errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("iso_download_pve can only be used together with iso_url"))
```

Comment (ELT4N1N) [reply builder/proxmox/common/config.go:369]: Yes it should be errs

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (nywilken) [inline builder/proxmox/common/step_download_iso_on_pve.go:36]: It wasn't immediately clear to me  what this was returning. I recommend documenting this function to indicate what it does.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (nywilken) [inline builder/proxmox/common/step_download_iso_on_pve.go:61]: Seeing as both checksum and checksumType are already empty strings as defined on lines 43 and 44. I recommend inverting this if/else statement to drop the unnecessary else.  
```suggestion
		if ISOChecksum != "none" {
			gr := &getter.Request{
				Src: url + "?checksum=" + ISOChecksum,
			}
			gc := getter.Client{}
			fileChecksum, err := gc.GetChecksum(context.TODO(), gr)
			if err != nil {
				state.Put("error", err)
				ui.Error(err.Error())
				continue
			}
			checksum = hex.EncodeToString(fileChecksum.Value)
			checksumType = fileChecksum.Type
		}
```

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (nywilken) [inline builder/proxmox/common/step_download_iso_on_pve.go:83]: Assuming if we it means no ISOUrls were specified, correct? 

If so I would recommend add a guard clause and return early before line 42

```go
if len(ISOUrls) == 0 {
 return ""
}
```

**Review by nywilken — COMMENTED** [review]
I left a few more suggestions and pinged @lbajolet-hashicorp for another set of eyes as he has more insight into the workings of Proxmox

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:36]: You are right. I will add some comments to make it more clear.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:61]: That's a good idea. I just took the code that was already there and didn't want to modify it too much. 
Didn't know how Go defines undefined strings. I'm not a Go developer :)

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:83]: It returns an empty string, if no download was successful and theoretically if no URLs were given (which should not be possible because of the check in the config).
If ISOUrls would be empty, the for loop should do nothing and therefore directly jump to return "".

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:81]: So in my understanding ISOUrls is an array of URLs to the same iso file (with the same checksum), basically mirror hosts. If the first URL is down or the file does not match the checksum it tries the next URL and so on. 
Therefore, if the first download was successful we don't need to try the other ones and can just return the path to the iso.

At least that's how I understood the code. I didn't try that feature yet because it seems to only be available for HCL2 configs and mine are JSON.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [reply builder/proxmox/common/step_download_iso_on_pve.go:44]: In terms of code style, Go encourages usage of camelCase or PascalCase for function names depending on visibility, so I would suggest we rename this to `DownloadISOOnPVE` in order to be more in-line with the guidelines for the language.

File: builder/proxmox/iso/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/iso/step_download_iso_on_pve.go:15]: Is there a reason to keep this step in the code since we're defining a `common` step for this purpose? I would think the `common` implementation supersedes this one, and therefore we could remove this code.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [reply builder/proxmox/common/step_download_iso_on_pve.go:93]: In case nothing was downloaded, this should be an error, shouldn't it? I think the function would benefit from returning both a string and an error with context on what went wrong, so we can forward this to the user, and they can troubleshoot more effectively.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/step_download_iso_on_pve.go:34]: Halting is correct here, but I would argue we should present an error to the user with the `Ui`. You can get this from the state bag, it's available under the `ui` key, and its type is `packersdk.Ui`.
SEE: builder/proxmox/iso/step_upload_iso.go:28 for an example of how to get it from the state.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:87]: This would benefit from documentation to make it clear what the behaviour should be if defined.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/step_download_iso_on_pve.go:83]: We should only set the error if all the attempts have failed, otherwise it's premature as subsequent tries may succeed.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/step_download_iso_on_pve.go:96]: I wonder, should we clean the downloaded ISOs from the server? Is there an API call to do this?

**Review by lbajolet-hashicorp — CHANGES_REQUESTED** [review]
Hi @ELT4N1N,

First off, apologies for the long turnaround time, I forgot to take a look at this when @nywilken mentioned me.

I took a look at the code, and left a couple of comments/questions which I think would benefit from being addressed before we merge this PR.

On those, one question that is in my opinion the most important is why do we have two separate steps for downloading those ISOs? I would think one step downloading all the ISOs would be enough for both the ISO for booting the image, and the other ISOs as well should there be more.

Other than that, the base code looks OK, there's room for improvement I think, but the framework for supporting this is in place, so now we can iterate on it.

Let me know if you need help on anything, I should be able to take a look at the code if need be.

Thanks for opening this PR!

**Comment by ELT4N1N** [comment]
Thanks for your feedback. I'm not an developer so your inputs are helpful to me. I thought by keeping most of the code that is already in packer it would be easier to merge. I will look into your suggestions.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:96]: I can look into it, but I would suggest keeping it behind a flag that enables or disables the cleanup. I don't think we have to delete the e. g. virtio drivers every time as it can be reused in the next template. At least an my use case of installing multiple windows versions it is faster when keeping the files.

File: builder/proxmox/common/config.go
Comment (ELT4N1N) [reply builder/proxmox/common/config.go:87]: added it to "clone" documentation as it already is in the "iso" part

File: builder/proxmox/iso/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/iso/step_download_iso_on_pve.go:16]: The problem is that the file path is returned in two different ways. In `common` I can return it via the `additionalISOsConfig` struct. But in `iso` I can only get a string variable, so it was solved by putting it in the `iso-config` state bag.
It is possible to move that logic to `stepUploadISO` and 'stepUploadAdditionalISO' and reuse the `DownloadISOOnPVE` function there. I don't know if this is a better solution.

File: builder/proxmox/common/step_download_iso_on_pve.go
Comment (ELT4N1N) [reply builder/proxmox/common/step_download_iso_on_pve.go:100]: It seems to be possible but I think that would be a new feature as it is also not done for the normal upload of iso files. The only iso file that gets deleted is the auto-generated one in `stepUploadAdditionalISO`. So I think this could be one solution for all these procedures and not specific to this step.

**Comment by ELT4N1N** [comment]
@lbajolet-hashicorp Sorry I overlooked your change request of the function renaming and now it seems to be a conflict that I can't solve form my side.

File: builder/proxmox/common/config.go
Comment (rgl) [reply builder/proxmox/common/config.go:87]: this variable name is somewhat ambiguous, can you please rename it to `iso_download_from_pve_node` (the `node` word is used to match the proxmox-iso source `node` property)?

**Comment by lbajolet-hashicorp** [comment]
Hey @ELT4N1N,

Sorry it took so long for us to take another look at this PR, let's get the ball rolling to get it merged soon!

I took the liberty to fix the conflicts on the code, it should be building on top of the current main now.

I'll take another look at the code now, if I can't finish my review today, you can expect it to be done by tomorrow.

Thanks!

**Pull request merged**
